datadog/serializer.go

```diff
@@ -0,0 +1,180 @@
+package datadog
+
+import (
+	"bytes"
+	"io"
+	"log"
+	"math"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/segmentio/stats/v4"
+)
+
+// Datagram format: https://docs.datadoghq.com/developers/dogstatsd/datagram_shell
+
+type serializer struct {
+	conn             net.Conn
+	bufferSize       int
+	filters          map[string]struct{}
+	distPrefixes     []string
+	useDistributions bool
+}
+
+func (s *serializer) Write(b []byte) (int, error) {
+	if s.conn == nil {
+		return 0, io.ErrClosedPipe
+	}
+
+	if len(b) <= s.bufferSize {
+		return s.conn.Write(b)
+	}
+
+	// When the serialized metrics are larger than the configured socket buffer
+	// size we split them on '\n' characters.
+	var n int
+
+	for len(b) != 0 {
+		var splitIndex int
+
+		for splitIndex != len(b) {
+			i := bytes.IndexByte(b[splitIndex:], '\n')
+			if i < 0 {
+				panic("stats/datadog: metrics are not formatted for the dogstatsd protocol")
+			}
+			if (i + splitIndex) >= s.bufferSize {
+				if splitIndex == 0 {
+					log.Printf("stats/datadog: metric of length %d B doesn't fit in the socket buffer of size %d B: %s", i+1, s.bufferSize, string(b))
+					b = b[i+1:]
+					continue
+				}
+				break
+			}
+			splitIndex += i + 1
+		}
+
+		c, err := s.conn.Write(b[:splitIndex])
+		if err != nil {
+			return n + c, err
+		}
+
+		n += c
+		b = b[splitIndex:]
+	}
+
+	return n, nil
+}
+
+func (s *serializer) close() {
```

Comment:
No error in the signature here? Seems better to return it and let the caller decide to ignore it.

Reply:
Ditto

These lines are green only because I moved them. I didn't actually write this code, and I'd like to keep this change as simple as possible by not editing the existing implementation.